In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
df=sns.load_dataset('iris')
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [ ]:
df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [ ]:
df['species'].unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [ ]:
df['species'].value_counts()

species
setosa        50
versicolor    50
virginica     50
Name: count, dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x=df.drop('species', axis=1)
y=df['species']

In [ ]:
x

sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

In [ ]:
x_train, x_test, y_train, y_test=train_test_split(x,y, test_size=0.2, random_state=42)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
model_knn=KNeighborsClassifier(n_neighbors=57) #We can change this n_neighbor like 3,5 ,25, and more. Each time I might get different accuracy
model_knn.fit(x_train, y_train)

KNeighborsClassifier(n_neighbors=57)

In [ ]:
model_knn.score(x_test, y_test)

0.9666666666666667

If you are getting accuracy score 100 % , model is overfit

SVM

In [ ]:
from sklearn.svm import SVC

In [ ]:
model_svm=SVC(C=30, kernel='rbf', gamma='auto')  #C is a punishment/penalty, agar decision boundary ke andar koi aaya
model_svm.fit(x_train, y_train)

SVC(C=30, gamma='auto')

In [ ]:
model_svm.score(x_test, y_test)

1.0

In [ ]:
#Applying cross validation in SVM model
from sklearn.model_selection import cross_val_score
scores=cross_val_score(model_svm, x,y, cv=6, scoring='accuracy')
scores

array([0.96, 1.  , 0.88, 0.92, 0.96, 1.  ])

Generally Each time we tweak/change parameters in models, we might get different accuracy. So how to identify which combination(grid) to use to get maximum accuracy, for that we use GRID SEARCH CV

GRID SEARCH CV Code Implementation

In [ ]:
from sklearn.model_selection import GridSearchCV

Performing HYPERPARAMETERTUNING through GRID Search CV

Grid Search CV for SVM Model

In [ ]:
classifier=GridSearchCV((model_svm),{
    'C': [1,10,20,30,50],
    'kernel': ['rbf', 'linear', 'poly'],
    'gamma': ['auto', 'scale']
}, cv=5, return_train_score=False) #Inside the dictionary there are multiple hyperparameters of the SVM model only

In [ ]:
classifier.fit(x,y)

GridSearchCV(cv=5, estimator=SVC(C=30, gamma='auto'),
             param_grid={'C': [1, 10, 20, 30, 50], 'gamma': ['auto', 'scale'],
                         'kernel': ['rbf', 'linear', 'poly']})

In [ ]:
results_gridcv_svm=pd.DataFrame(classifier.cv_results_)
results_gridcv_svm

#This will show the different cobinations(grid) of the hyperparameters ,
#which you gave while creating SVM model in Grid Search CV inside the dictionary(in above code)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  param_C  \
0        0.010031      0.005202         0.005475        0.002058        1   
1        0.003663      0.000160         0.003516        0.000866        1   
2        0.007250      0.002623         0.002854        0.000070        1   
3        0.003841      0.000268         0.002889        0.000172        1   
4        0.004175      0.000696         0.002754        0.000118        1   
5        0.005302      0.001894         0.003970        0.001592        1   
6        0.004027      0.000637         0.003509        0.001717       10   
7        0.006866      0.002288         0.005055        0.003371       10   
8        0.009373      0.004747         0.003867        0.002038       10   
9        0.003362      0.000102         0.002564        0.000205       10   
10       0.006663      0.002852         0.004166        0.001997       10   
11       0.004253      0.000823         0.003562        0.001710       10   
12       0.003539      0.000187         0.002783        0.000180       20   
13       0.003433      0.000080         0.002529        0.000105       20   
14       0.023040      0.019519         0.004557        0.003012       20   
15       0.004627      0.001588         0.003249        0.000715       20   
16       0.005873      0.002220         0.003371        0.000957       20   
17       0.007599      0.003194         0.005261        0.002516       20   
18       0.004003      0.000278         0.003547        0.001560       30   
19       0.004620      0.001611         0.003531        0.001391       30   
20       0.019654      0.012820         0.003138        0.000342       30   
21       0.004411      0.001489         0.003554        0.001615       30   
22       0.004041      0.000715         0.002641        0.000078       30   
23       0.004163      0.000068         0.002871        0.000212       30   
24       0.003704      0.000159         0.002729        0.000107       50   
25       0.003538      0.000116         0.003038        0.000462       50   
26       0.020355      0.011988         0.003892        0.001377       50   
27       0.004156      0.000397         0.003037        0.000267       50   
28       0.004560      0.001238         0.002975        0.000267       50   
29       0.007218      0.003704         0.004658        0.002239       50   

   param_gamma param_kernel                                           params  \
0         auto          rbf       {'C': 1, 'gamma': 'auto', 'kernel': 'rbf'}   
1         auto       linear    {'C': 1, 'gamma': 'auto', 'kernel': 'linear'}   
2         auto         poly      {'C': 1, 'gamma': 'auto', 'kernel': 'poly'}   
3        scale          rbf      {'C': 1, 'gamma': 'scale', 'kernel': 'rbf'}   
4        scale       linear   {'C': 1, 'gamma': 'scale', 'kernel': 'linear'}   
5        scale         poly     {'C': 1, 'gamma': 'scale', 'kernel': 'poly'}   
6         auto          rbf      {'C': 10, 'gamma': 'auto', 'kernel': 'rbf'}   
7         auto       linear   {'C': 10, 'gamma': 'auto', 'kernel': 'linear'}   
8         auto         poly     {'C': 10, 'gamma': 'auto', 'kernel': 'poly'}   
9        scale          rbf     {'C': 10, 'gamma': 'scale', 'kernel': 'rbf'}   
10       scale       linear  {'C': 10, 'gamma': 'scale', 'kernel': 'linear'}   
11       scale         poly    {'C': 10, 'gamma': 'scale', 'kernel': 'poly'}   
12        auto          rbf      {'C': 20, 'gamma': 'auto', 'kernel': 'rbf'}   
13        auto       linear   {'C': 20, 'gamma': 'auto', 'kernel': 'linear'}   
14        auto         poly     {'C': 20, 'gamma': 'auto', 'kernel': 'poly'}   
15       scale          rbf     {'C': 20, 'gamma': 'scale', 'kernel': 'rbf'}   
16       scale       linear  {'C': 20, 'gamma': 'scale', 'kernel': 'linear'}   
17       scale         poly    {'C': 20, 'gamma': 'scale', 'kernel': 'poly'}   
18        auto          rbf      {'C': 30, 'gamma': 'auto', 'kernel': 'rbf'}   
19        auto  

In [ ]:
results_gridcv_svm.columns

Index(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time',
       'param_C', 'param_gamma', 'param_kernel', 'params', 'split0_test_score',
       'split1_test_score', 'split2_test_score', 'split3_test_score',
       'split4_test_score', 'mean_test_score', 'std_test_score',
       'rank_test_score'],
      dtype='object')

In [ ]:
#The above output result is too big , but I am only interested in 'param_C', 'param_gamma', 'param_kernel, 'mean_test_score' i.e, accuracy

results_gridcv_svm[['param_C', 'param_gamma', 'param_kernel', 'mean_test_score']]


param_C param_gamma param_kernel  mean_test_score
0         1        auto          rbf         0.980000
1         1        auto       linear         0.980000
2         1        auto         poly         0.966667
3         1       scale          rbf         0.966667
4         1       scale       linear         0.980000
5         1       scale         poly         0.980000
6        10        auto          rbf         0.980000
7        10        auto       linear         0.973333
8        10        auto         poly         0.966667
9        10       scale          rbf         0.980000
10       10       scale       linear         0.973333
11       10       scale         poly         0.966667
12       20        auto          rbf         0.966667
13       20        auto       linear         0.966667
14       20        auto         poly         0.953333
15       20       scale          rbf         0.980000
16       20       scale       linear         0.966667
17       20       scale         poly         0.966667
18       30        auto          rbf         0.960000
19       30        auto       linear         0.960000
20       30        auto         poly         0.953333
21       30       scale          rbf         0.973333
22       30       scale       linear         0.960000
23       30       scale         poly         0.966667
24       50        auto          rbf         0.960000
25       50        auto       linear         0.966667
26       50        auto         poly         0.960000
27       50       scale          rbf         0.973333
28       50       scale       linear         0.966667
29       50       scale         poly         0.966667

So from above output we conlclude ,while creating SVM model  if I put (C=1, kernel=rbf , gamma= auto), I will get hifgest accuracy 98%

GRID SEARCH CV for kNN Model

In [ ]:
classifier2=GridSearchCV((model_knn),{
    'n_neighbors': [1,3,5,7,9],
    'weights': ['distance', 'uniform'],
    'algorithm' : ['auto', 'ball_tree', 'kd_tree', 'brute'],
}, cv=5 )

In [ ]:
classifier2.fit(x,y)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(n_neighbors=57),
             param_grid={'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
                         'n_neighbors': [1, 3, 5, 7, 9],
                         'weights': ['distance', 'uniform']})

In [ ]:
results_gridcv_knn=pd.DataFrame(classifier2.cv_results_)
results_gridcv_knn

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.005459      0.003868         0.004108        0.000500   
1        0.003527      0.000064         0.006490        0.003047   
2        0.003908      0.000950         0.004455        0.000809   
3        0.003810      0.000204         0.005827        0.001913   
4        0.003354      0.000054         0.004587        0.001092   
5        0.006596      0.002571         0.007984        0.002871   
6        0.007704      0.003747         0.008049        0.005034   
7        0.004042      0.000674         0.005772        0.001280   
8        0.003273      0.000078         0.003673        0.000067   
9        0.003493      0.000171         0.004742        0.000183   
10       0.008793      0.009599         0.007779        0.003344   
11       0.007893      0.005536         0.012607        0.014028   
12       0.004172      0.001217         0.005151        0.001821   
13       0.003418      0.000239         0.004730        0.000377   
14       0.003794      0.000541         0.008423        0.004944   
15       0.009767      0.005108         0.013335        0.002983   
16       0.004840      0.001595         0.005428        0.002208   
17       0.003955      0.000936         0.005961        0.001725   
18       0.004551      0.002564         0.004494        0.001388   
19       0.003573      0.000328         0.005454        0.001504   
20       0.006245      0.002024         0.006477        0.002731   
21       0.004400      0.001667         0.005428        0.001168   
22       0.003422      0.000073         0.003703        0.000085   
23       0.003653      0.000346         0.004695        0.000093   
24       0.003671      0.000309         0.004679        0.001268   
25       0.003425      0.000209         0.004752        0.000228   
26       0.003525      0.000230         0.003770        0.000091   
27       0.004433      0.002089         0.007494        0.003407   
28       0.010799      0.005279         0.011276        0.004424   
29       0.008911      0.003251         0.011790        0.003917   
30       0.006081      0.002667         0.017218        0.025166   
31       0.011563      0.005882         0.011443        0.001275   
32       0.009034      0.004610         0.008347        0.003544   
33       0.005712      0.002292         0.006937        0.002565   
34       0.003070      0.000276         0.003611        0.000248   
35       0.006336      0.004799         0.008199        0.004332   
36       0.003471      0.000615         0.004569        0.001750   
37       0.002990      0.000097         0.004963        0.001057   
38       0.002981      0.000153         0.003429        0.000048   
39       0.003015      0.000059         0.004555        0.000070   

   param_algorithm  param_n_neighbors param_weights  \
0             auto                  1      distance   
1             auto                  1       uniform   
2             auto                  3      distance   
3             auto                  3       uniform   
4             auto                  5      distance   
5             auto                  5       uniform   
6             auto                  7      distance   
7             auto                  7       uniform   
8             auto                  9      distance   
9             auto                  9       uniform   
10       ball_tree                  1      distance   
11       ball_tree                  1       uniform   
12       ball_tree                  3      distance   
13       ball_tree                  3       uniform   
14       ball_tree                  5      distance   
15       ball_tree                  5       uniform   
16       ball_tree                  7      distance   
17       ball_tree                  7       uniform   
18       ball_tree                  9      distance   
19       ball_tree                  9       uniform   
20         kd_tree                  1      distance   
21   

In [ ]:
results_gridcv_knn.columns

Index(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time',
       'param_algorithm', 'param_n_neighbors', 'param_weights', 'params',
       'split0_test_score', 'split1_test_score', 'split2_test_score',
       'split3_test_score', 'split4_test_score', 'mean_test_score',
       'std_test_score', 'rank_test_score'],
      dtype='object')

In [ ]:
results_gridcv_knn[['param_n_neighbors','param_algorithm', 'param_weights','mean_test_score' ]]

param_n_neighbors param_algorithm param_weights  mean_test_score
0                   1            auto      distance         0.960000
1                   1            auto       uniform         0.960000
2                   3            auto      distance         0.966667
3                   3            auto       uniform         0.966667
4                   5            auto      distance         0.966667
5                   5            auto       uniform         0.973333
6                   7            auto      distance         0.980000
7                   7            auto       uniform         0.980000
8                   9            auto      distance         0.973333
9                   9            auto       uniform         0.973333
10                  1       ball_tree      distance         0.960000
11                  1       ball_tree       uniform         0.960000
12                  3       ball_tree      distance         0.966667
13                  3       ball_tree       uniform         0.966667
14                  5       ball_tree      distance         0.966667
15                  5       ball_tree       uniform         0.973333
16                  7       ball_tree      distance         0.980000
17                  7       ball_tree       uniform         0.980000
18                  9       ball_tree      distance         0.973333
19                  9       ball_tree       uniform         0.973333
20                  1         kd_tree      distance         0.960000
21                  1         kd_tree       uniform         0.960000
22                  3         kd_tree      distance         0.966667
23                  3         kd_tree       uniform         0.966667
24                  5         kd_tree      distance         0.966667
25                  5         kd_tree       uniform         0.973333
26                  7         kd_tree      distance         0.980000
27                  7         kd_tree       uniform         0.980000
28                  9         kd_tree      distance         0.973333
29                  9         kd_tree       uniform         0.973333
30                  1           brute      distance         0.960000
31                  1           brute       uniform         0.960000
32                  3           brute      distance         0.966667
33                  3           brute       uniform         0.966667
34                  5           brute      distance         0.966667
35                  5           brute       uniform         0.973333
36                  7           brute      distance         0.980000
37                  7           brute       uniform         0.980000
38                  9           brute      distance         0.973333
39                  9           brute       uniform         0.980000

In [ ]:
from sklearn.model_selection import RandomizedSearchCV

RANDOM SEARCH CV

Random Search CV for SVM Model

In [ ]:
classifier_random=RandomizedSearchCV((model_svm), {
    'C': [1,10,20,30,50],
    'kernel': ['rbf', 'linear', 'poly'],
    'gamma': ['auto', 'scale']
},n_iter=10, cv=5, return_train_score=True)
#Random Serach CV is same as Grid Search CV, the only thing added in random cv is n_iter =10, means it will select 10 model parameters randomly
#out of all possible combination i.e, 30 in grid search cv above


In [ ]:
classifier_random.fit(x,y)

RandomizedSearchCV(cv=5, estimator=SVC(C=30, gamma='auto'),
                   param_distributions={'C': [1, 10, 20, 30, 50],
                                        'gamma': ['auto', 'scale'],
                                        'kernel': ['rbf', 'linear', 'poly']},
                   return_train_score=True)

In [ ]:
#C=10 in above output , beacuse he randomly picks it , and found best

In [ ]:
results_random=pd.DataFrame(classifier_random.cv_results_)
results_random
#The output will show only 10 models /combination(grid) of parameters, as we input 10 random n_iterator while creating model

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_kernel  \
0       0.006285      0.002811         0.002005        0.000182         poly   
1       0.002520      0.000250         0.001732        0.000095       linear   
2       0.002507      0.000302         0.001724        0.000087          rbf   
3       0.002397      0.000111         0.001661        0.000062          rbf   
4       0.002202      0.000052         0.001572        0.000023       linear   
5       0.002315      0.000042         0.001694        0.000125         poly   
6       0.002196      0.000073         0.001715        0.000303       linear   
7       0.002484      0.000067         0.001727        0.000084          rbf   
8       0.002328      0.000078         0.001659        0.000079       linear   
9       0.002299      0.000034         0.001647        0.000071          rbf   

  param_gamma  param_C                                           params  \
0        auto       10     {'kernel': 'poly', 'gamma': 'auto', 'C': 10}   
1       scale       20  {'kernel': 'linear', 'gamma': 'scale', 'C': 20}   
2       scale       50     {'kernel': 'rbf', 'gamma': 'scale', 'C': 50}   
3        auto       10      {'kernel': 'rbf', 'gamma': 'auto', 'C': 10}   
4        auto       10   {'kernel': 'linear', 'gamma': 'auto', 'C': 10}   
5       scale       10    {'kernel': 'poly', 'gamma': 'scale', 'C': 10}   
6        auto       50   {'kernel': 'linear', 'gamma': 'auto', 'C': 50}   
7       scale        1      {'kernel': 'rbf', 'gamma': 'scale', 'C': 1}   
8       scale       50  {'kernel': 'linear', 'gamma': 'scale', 'C': 50}   
9        auto       50      {'kernel': 'rbf', 'gamma': 'auto', 'C': 50}   

   split0_test_score  split1_test_score  ...  mean_test_score  std_test_score  \
0           1.000000           1.000000  ...         0.966667        0.042164   
1           1.000000           1.000000  ...         0.966667        0.042164   
2           0.966667           1.000000  ...         0.973333        0.024944   
3           0.966667           1.000000  ...         0.980000        0.016330   
4           1.000000           1.000000  ...         0.973333        0.038873   
5           1.000000           1.000000  ...         0.966667        0.042164   
6           1.000000           1.000000  ...         0.966667        0.042164   
7           0.966667           0.966667  ...         0.966667        0.021082   
8           1.000000           1.000000  ...         0.966667        0.042164   
9           1.000000           0.966667  ...         0.960000        0.038873   

   rank_test_score  split0_train_score  split1_train_score  \
0                4            0.983333            0.983333   
1                4            0.975000            0.966667   
2                2            0.966667            0.966667   
3                1            0.975000            0.966667   
4                2            0.966667            0.966667   
5                4            0.975000            0.975000   
6                4            0.975000            0.975000   
7                4            0.983333            0.958333   
8                4            0.975000            0.975000   
9               10            0.983333            0.975000   

   split2_train_score  split3_train_score  split4_train_score  \
0            0.991667            1.000000            0.983333   
1            0.991667            1.000000            0.975000   
2            0.991667            0.991667            0.966667   
3            0.991667            0.983333            0.966667   
4            0.991667            0.991667            0.975000   
5            0.991667            0.991667            0.975000   
6            0.991667            1.000000            0.975000   
7            0.983333            0.983333            0.958333   
8            0.991667            1.000000            0.975000   
9            0.991667            1.000000            0.983333   

   me

In [ ]:
results_random.columns

Index(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time',
       'param_kernel', 'param_gamma', 'param_C', 'params', 'split0_test_score',
       'split1_test_score', 'split2_test_score', 'split3_test_score',
       'split4_test_score', 'mean_test_score', 'std_test_score',
       'rank_test_score', 'split0_train_score', 'split1_train_score',
       'split2_train_score', 'split3_train_score', 'split4_train_score',
       'mean_train_score', 'std_train_score'],
      dtype='object')

In [ ]:
results_random[['param_C', 'param_gamma', 'param_kernel', 'mean_test_score']]

param_C param_gamma param_kernel  mean_test_score
0       10        auto         poly         0.966667
1       20       scale       linear         0.966667
2       50       scale          rbf         0.973333
3       10        auto          rbf         0.980000
4       10        auto       linear         0.973333
5       10       scale         poly         0.966667
6       50        auto       linear         0.966667
7        1       scale          rbf         0.966667
8       50       scale       linear         0.966667
9       50        auto          rbf         0.960000

Best perfornming model above is c=10, kernel='rbf', gamma='auto'-- accuracy have 98%

RANDOM SEARCH CV for kNN Model

In [ ]:
classifier_random_knn=RandomizedSearchCV((model_knn),{
    'n_neighbors': [1,3,5,7,9],
    'weights': ['distance', 'uniform'],
    'algorithm' : ['auto', 'ball_tree', 'kd_tree', 'brute'],
}, n_iter=10, cv=5, return_train_score=False)

In [ ]:
classifier_random_knn.fit(x,y)

RandomizedSearchCV(cv=5, estimator=KNeighborsClassifier(n_neighbors=57),
                   param_distributions={'algorithm': ['auto', 'ball_tree',
                                                      'kd_tree', 'brute'],
                                        'n_neighbors': [1, 3, 5, 7, 9],
                                        'weights': ['distance', 'uniform']})

In [ ]:
results_random_knn=pd.DataFrame(classifier_random_knn.cv_results_)
results_random_knn

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_weights  \
0       0.007549      0.003081         0.009114        0.001717       uniform   
1       0.006685      0.001996         0.011103        0.002400       uniform   
2       0.016719      0.008680         0.025386        0.012959       uniform   
3       0.023060      0.005370         0.031346        0.005252       uniform   
4       0.028414      0.004756         0.044029        0.006428       uniform   
5       0.026937      0.008210         0.027741        0.007031      distance   
6       0.011486      0.008647         0.019469        0.009397       uniform   
7       0.006828      0.001104         0.007970        0.000072      distance   
8       0.008695      0.003023         0.012042        0.004435       uniform   
9       0.010672      0.002076         0.011856        0.003543      distance   

   param_n_neighbors param_algorithm  \
0                  3       ball_tree   
1                  7           brute   
2                  1       ball_tree   
3                  3           brute   
4                  1           brute   
5                  3            auto   
6                  5           brute   
7                  9       ball_tree   
8                  1            auto   
9                  9         kd_tree   

                                              params  split0_test_score  \
0  {'weights': 'uniform', 'n_neighbors': 3, 'algo...           0.966667   
1  {'weights': 'uniform', 'n_neighbors': 7, 'algo...           0.966667   
2  {'weights': 'uniform', 'n_neighbors': 1, 'algo...           0.966667   
3  {'weights': 'uniform', 'n_neighbors': 3, 'algo...           0.966667   
4  {'weights': 'uniform', 'n_neighbors': 1, 'algo...           0.966667   
5  {'weights': 'distance', 'n_neighbors': 3, 'alg...           0.966667   
6  {'weights': 'uniform', 'n_neighbors': 5, 'algo...           0.966667   
7  {'weights': 'distance', 'n_neighbors': 9, 'alg...           0.966667   
8  {'weights': 'uniform', 'n_neighbors': 1, 'algo...           0.966667   
9  {'weights': 'distance', 'n_neighbors': 9, 'alg...           0.966667   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.966667           0.933333           0.966667                1.0   
1           1.000000           0.966667           0.966667                1.0   
2           0.966667           0.933333           0.933333                1.0   
3           0.966667           0.933333           0.966667                1.0   
4           0.966667           0.933333           0.933333                1.0   
5           0.966667           0.933333           0.966667                1.0   
6           1.000000           0.933333           0.966667                1.0   
7           1.000000           0.933333           0.966667                1.0   
8           0.966667           0.933333           0.933333                1.0   
9           1.000000           0.933333           0.966667                1.0   

   mean_test_score  std_test_score  rank_test_score  
0         0.966667        0.021082                5  
1         0.980000        0.016330                1  
2         0.960000        0.024944                8  
3         0.966667        0.021082                5  
4         0.960000        0.024944                8  
5         0.966667        0.021082                5  
6         0.973333        0.024944                2  
7         0.973333        0.024944                2  
8         0.960000        0.024944                8  
9         0.973333        0.024944                2

In [ ]:
results_random_knn.columns

Index(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time',
       'param_weights', 'param_n_neighbors', 'param_algorithm', 'params',
       'split0_test_score', 'split1_test_score', 'split2_test_score',
       'split3_test_score', 'split4_test_score', 'mean_test_score',
       'std_test_score', 'rank_test_score'],
      dtype='object')

In [ ]:
results_random_knn[['param_n_neighbors','param_algorithm', 'param_weights','mean_test_score' ]]

param_n_neighbors param_algorithm param_weights  mean_test_score
0                  3       ball_tree       uniform         0.966667
1                  7           brute       uniform         0.980000
2                  1       ball_tree       uniform         0.960000
3                  3           brute       uniform         0.966667
4                  1           brute       uniform         0.960000
5                  3            auto      distance         0.966667
6                  5           brute       uniform         0.973333
7                  9       ball_tree      distance         0.973333
8                  1            auto       uniform         0.960000
9                  9         kd_tree      distance         0.973333

Best performing model above is n_neighbors=7, algorithm='brute', weights='uniform'; accuracy having 98%

What Actually Happens i n RANDOM SEARCH CV model:
Randomly selects 10 parameter combinations from the 40 possible ones

For each of these 10 selected combinations, it does 5-fold CV

So it trains: 10 combinations × 5 folds = 50 models total

Random Selection: Randomly picks 10 hyperparameter combinations from all possibilities

Cross-Validation: For each combination, performs 5-fold CV

Scoring: Evaluates performance on validation folds

Best Model: Returns the combination with highest average validation score

Missing Parameters (Usually Default)
Scoring metric: Not specified, uses classifier's default (usually accuracy for classification)

Random state: Not set, so results may vary between runs

n_jobs: Not specified, uses single processor

Usage Example
python
# After running the search
classifier_random_knn.fit(X_train, y_train)

# Get best parameters
print("Best parameters:", classifier_random_knn.best_params_)
print("Best score:", classifier_random_knn.best_score_)

# Use best model for predictions
best_knn = classifier_random_knn.best_estimator_
predictions = best_knn.predict(X_test)
Advantages Over GridSearchCV
Faster: Tests only a subset of parameters

Better for large spaces: More efficient when many hyperparameters

Can find good combinations: Often finds near-optimal parameters quickly

This approach balances computational efficiency with finding good hyperparameter values for your KNN model.

what is return_test_score, explain in simple way
What is return_train_score? (Simple Explanation)
Think of it like this: When you're practicing for an exam...

Training score = Your score on practice questions (the ones you studied)

Test/Validation score = Your score on real exam questions (new, unseen questions)

The Setting:
python
return_train_score=False
What This Means:
return_train_score=False = "Don't show me my practice scores, just show me my exam scores"

Simple Example:
Let's say you're tuning your KNN model with different numbers of neighbors:

n_neighbors	Practice Score	Exam Score
1	98%	75%
3	95%	82%
5	92%	85%
With return_train_score=False, the code only remembers:

Best model: n_neighbors=5 with 85% exam score

Why Turn It Off?
Saves Memory: Practice scores take up space to remember

Faster: Less data to process and store

Focus on What Matters: Only care about how well you'll do on NEW questions (the exam), not the ones you already practiced

When You Might Want It On:
python
return_train_score=True  # Show me both practice AND exam scores
Use this when you want to check for overfitting:

If practice score = 98% but exam score = 75% → You memorized instead of learning!

This helps you understand if your model is too complex

Bottom Line:
return_train_score=False = "Just tell me which settings work best on new data, don't bother remembering how they did on the training data."

It's like saying: "I don't care how well I memorized the practice questions - I only care about how I'll perform on the actual test!"

okay there are 40 total models and each model there is k=5 , it means there are 40*5 models training in this simple code, and for each model out of 40 , there comes average of 5 models, and finally it choose randomly 10 models out of 40 . Am I right?
Almost perfect! You're very close, but let me clarify one important detail about the random selection.

Your Understanding - Mostly Correct! ✅
You got these parts right:

✅ 40 total possible combinations (5 × 2 × 4 parameter combinations)

✅ Each combination gets k=5 folds in cross-validation

✅ For each of the 40 combinations, you get an average of 5 models' performance

The One Correction: 🔄
"choose randomly 10 models out of 40" - This is the part that needs clarification.

It's not choosing 10 completed models from 40, but rather:

What Actually Happens:
Randomly selects 10 parameter combinations from the 40 possible ones

For each of these 10 selected combinations, it does 5-fold CV

So it trains: 10 combinations × 5 folds = 50 models total

Visual Example:
All Possible Combinations (40):

text
Combination 1: n_neighbors=1, weights='uniform', algorithm='auto'
Combination 2: n_neighbors=1, weights='uniform', algorithm='ball_tree'
...
Combination 40: n_neighbors=9, weights='distance', algorithm='brute'
RandomizedSearchCV with n_iter=10:

Randomly picks 10 of these combinations (say, #3, #7, #12, #15, #18, #22, #25, #31, #35, #39)

For each of these 10, trains 5 models (one per fold)

Total models trained: 10 × 5 = 50

Comparison:
Method	Parameter Combinations Tested	Total Models Trained
GridSearchCV	All 40 combinations	40 × 5 = 200 models
RandomizedSearchCV (n_iter=10)	Random 10 combinations	10 × 5 = 50 models
So You Were Close!
You understood the cross-validation part perfectly! The only adjustment is that the random selection happens BEFORE training, not after.

Your mental model: "Choose randomly 10 from 40, then train each with 5-fold CV" ✅

Great thinking! You clearly understand the core concepts of hyperparameter tuning and cross-validation.

i did not understand , Explain me  more simple way like I am 10 year old
Let's Imagine You're Tasting Cookies! 🍪
You have 40 different cookie recipes to try:

5 types of chocolate chips

2 types of sugar

4 types of flour

If you were GridSearchCV (the thorough method):
You would bake ALL 40 recipes

Taste each recipe 5 times to be sure

Total cookies baked = 40 × 5 = 200 cookies! 🍪🍪🍪

This takes FOREVER!

But you're RandomizedSearchCV (the smart method):
Your mom says: "Only bake 10 random recipes from the 40"

So you randomly pick 10 recipes (like lucky dip!)

For each of these 10 recipes, you bake it 5 times to be sure it's good

Total cookies baked = 10 × 5 = 50 cookies 🍪

Much faster!

What the Code Does:
python
n_iter=10  # "Only try 10 random recipes"
cv=5       # "Taste each recipe 5 times to be sure"
So You Were ALMOST Right!
You thought: "Try all 40 recipes, then pick 10 best"
Actually: "Pick 10 random recipes first, then only try those"

Simple Summary:
40 recipes = all possible combinations

n_iter=10 = only bake 10 random recipes

cv=5 = taste each chosen recipe 5 times

Total = 10 recipes × 5 tastings = 50 total tastings

You save time by not baking all 40 recipes! 🎯

if there are total 50 , then in output why its giving 10 only
Great question! Let me explain this cookie mystery! 🍪

What's Happening:
You baked 10 recipes, but tasted each recipe 5 times

Cookie Example:
You have 10 cookie recipes you randomly picked

For each recipe, you ask 5 friends to taste it and give a score

So: 10 recipes × 5 friends = 50 total tastings

But When You Report Results:
You don't say "Friend 1 gave Recipe A: 8/10, Friend 2 gave Recipe A: 7/10..."

You say: "Recipe A got 7.5/10 on average"

You combine the 5 scores into 1 average score per recipe

In Code Terms:
python
# What happens internally:
Recipe 1: Scores [0.85, 0.82, 0.83, 0.84, 0.81] → Average: 0.83
Recipe 2: Scores [0.78, 0.79, 0.80, 0.77, 0.79] → Average: 0.79
...
Recipe 10: Scores [0.90, 0.88, 0.89, 0.91, 0.87] → Average: 0.89
The Output Shows:
Only 10 final results - because it's showing you:

10 recipes (parameter combinations)

Each with its 1 average score (from 5 tastings/folds)

What You See vs What Happens:
What Happens Behind the Scenes	What You See in Output
50 model trainings (10 recipes × 5 tastings)	10 average scores (1 per recipe)
Lots of work!	Clean, simple results
So You're Right to Be Confused!
You're thinking: "I did 50 trainings, show me 50 results!"
But the code is smart - it combines the 5 scores into 1 average so you can easily compare the recipes! 🎯

Bottom line: 50 trainings happened, but you only see 10 summarized results - one for each parameter combination!

Okay explaini me return_train_score=False/True  with this same example
Let's Use the Cookie Example! 🍪
Remember:

10 cookie recipes (parameter combinations)

5 friends taste each recipe (5-fold CV)

Each friend gives a score from 1-10

return_train_score=False ❌
What happens:
You only ask friends: "How do these cookies taste?" (test scores)

You DON'T ask: "How was the baking process?" (train scores)

Results you get:
text
Recipe 1: Average taste score = 8.2/10
Recipe 2: Average taste score = 7.8/10
...
Recipe 10: Average taste score = 9.1/10  ← WINNER!
Only 10 numbers - just the final taste scores!

return_train_score=True ✅
What happens:
You ask TWO questions for each recipe:

"How was the BAKING process?" (train score)

Did the ingredients mix well?

Was the dough easy to work with?

"How do the COOKIES TASTE?" (test score)

The final result

Results you get:
text
Recipe 1:
- Baking score = 9.8/10  (train_score)
- Taste score = 8.2/10   (test_score)

Recipe 2:
- Baking score = 9.5/10  (train_score)  
- Taste score = 7.8/10   (test_score)

...
Recipe 10:
- Baking score = 9.9/10  (train_score)
- Taste score = 9.1/10   (test_score)  ← WINNER!
Now you get 20 numbers - 10 baking scores + 10 taste scores!

Why This Matters:
If Baking Score = 10/10 but Taste Score = 5/10:
Means: You followed the recipe perfectly but cookies taste bad!

In ML terms: The model memorized the training data but can't handle new data (Overfitting)

return_train_score=False is like:
"I don't care about the baking process - just tell me which cookies taste best!" 🎯

return_train_score=True is like:
"Show me both - I want to see if any recipes look good during baking but taste bad later!" 🔍

Simple Summary:
False = Only final exam scores

True = Both practice test scores + final exam scores

Most of the time, we use False because we only care about which model works best on new data! 🍪✅